Sources/Target/Linux/Thread.cpp

```diff
@@ -35,36 +35,16 @@ namespace Linux {
 
 Thread::Thread(Process *process, ThreadId tid) : super(process, tid) {}
 
-void Thread::fillWatchpointData() {
+void Thread::setBreakpointStopReason() {
   HardwareBreakpointManager *hwBpm = process()->hardwareBreakpointManager();
-
-  if (hwBpm) {
+  if (hwBpm != nullptr) {
     BreakpointManager::Site site;
-
     int bpIdx = hwBpm->hit(this, site);
     if (bpIdx >= 0) {
-      _stopInfo.watchpointIndex = bpIdx;
-      _stopInfo.watchpointAddress = site.address;
-
-      switch (static_cast<int>(site.mode)) {
-      case BreakpointManager::kModeExec:
-        _stopInfo.reason = StopInfo::kReasonBreakpoint;
-        return;
-      case BreakpointManager::kModeWrite:
-        _stopInfo.reason = StopInfo::kReasonWriteWatchpoint;
-        return;
-      case BreakpointManager::kModeRead:
-        _stopInfo.reason = StopInfo::kReasonReadWatchpoint;
-        return;
-      case BreakpointManager::kModeRead | BreakpointManager::kModeWrite:
-        _stopInfo.reason = StopInfo::kReasonAccessWatchpoint;
-        return;
-      default:
-        DS2BUG("invalid mode");
-      }
+      setHardwareBreakpointStopReason(bpIdx, site);
+      return;
     }
   }
-
   _stopInfo.reason = StopInfo::kReasonTrace;
 }
```

Comment:
Hoist and early return?

    _stopInfo.reason = StopInfo::kReasonTrace;
    auto *manager = process()->hardwareBreakpointManager();
    if (manager == nullptr)
      return;

    BreakpointManager::Site site;
    int index = manager->hit(this, site);
    if (index < 0)
      return;

    setHardwareBreakpointStopReason(index, site);

Reply:
Mmm, good idea.